### PR TITLE
Codecs Derivation

### PR DESCRIPTION
**Pull request by tabdulradi**
Supports Options, Seq, Set, and Map without Macros

### CODE REVIEW

File: troy-driver/src/main/scala/troy/driver/codecs/AdapterTypeCodec.scala [hunk 1]
```diff
@@ -20,4 +20,12 @@ abstract class AdapterTypeCodec[In, Out](innerCodec: TypeCodec[In], javaClass: C
 
   override def deserialize(bytes: ByteBuffer, protocolVersion: ProtocolVersion): Out =
     innerToOuter(innerCodec.deserialize(bytes, protocolVersion))
+}
+
+class OptionTypeCodec[T <: AnyRef](innerCodec: TypeCodec[T]) extends AdapterTypeCodec[T, Option[T]](innerCodec, classOf[Option[T]]) {
+  private var empty: T = _
```
Comment (tabdulradi) [inline troy-driver/src/main/scala/troy/driver/codecs/AdapterTypeCodec.scala]: خليك في حالك

File: troy-driver/src/test/scala/troy/driver/Usage.scala [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+object Usage extends App {
+  import DSL._
+
+  val cluster = Cluster.builder().addContactPoint("127.0.0.1").build()
+  implicit val session: Session = cluster.connect()
+  case class Post(id: UUID, author_name: String, title: String)
+  import scala.concurrent.ExecutionContext.Implicits.global
+
+  val getByTitle = {
+    import InternalDsl._
```
Comment (tabdulradi) [inline troy-driver/src/test/scala/troy/driver/Usage.scala]: This is what the macro generates, not changing this.

File: troy-driver/src/test/scala/troy/driver/CodecsSpec.scala [hunk 3]
```diff
@@ -0,0 +1,1710 @@
...
+
+        "get from Row" in {
+          implicit val gettable = FakeAbstractGettableByIndexData.timestamp(jValue)
+          column[SType](0).as[CType] mustBe sValue
+        }
+        "set into BoundStatement" in {
+          param(sValue).as[CType].set(settable, 0)
+        }
+        "get from Row (nullable)" in {
+          implicit val gettable = FakeAbstractGettableByIndexData.timestamp(null)
```
Comment (tabdulradi) [inline troy-driver/src/test/scala/troy/driver/CodecsSpec.scala]: This code is interacting with Java code, I need to pass null.

**Comment by tabdulradi** [comment]
@SalmaKhater The PR is ready for your review.

File: troy-driver/src/test/scala/troy/driver/Usage.scala [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        )
+
+      bind(prepared, param(title).as[CDT.Text])
+        .executeAsync
+        .oneOption
+        .parseAs(parser)
+    }
+  }
+
+  println(Await.result(getByTitle("Title"), Duration(1, "second")))
```
Comment (SalmaKhater) [inline troy-driver/src/test/scala/troy/driver/Usage.scala]: why println ?!

**Pull request merged**